---
title: "theodora phyloseq rarefaction"
output: html_document
date: "2024-05-20"
---

```{r phyloseq setup}
library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(tidyverse)
library(here)
library(dplyr)
library(vegan)
library(here)

# Set plot theme to a simple black-and-white design
theme_set(theme_bw())
```

```{r phyloseq preprocessing}
# Import output files from DADA2 (wide format asv). 
asvs <- read.csv(here("data", "sequencing_data", "Baja_MiFish_ASVs_wide.csv"),row.names=1)
tax <- read.csv (here("data", "sequencing_data", "Baja_MiFish_taxa_table.csv"))
hash <- read.csv(here("data", "sequencing_data","Baja_MiFish_hash_key.csv"))

# NASTASSIA'S FILES: Convert tax so that each level of taxonomy is its own column
# tax <- tax %>%
#   separate(Taxon, into = c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), sep = ";", fill = "right")
# 
# colnames(tax)[1] <- "Hash"

# Add sequences to tax
# tax_updated <- tax %>%
#   left_join(hash, tax, by = "Hash") %>%
#   select(-Hash.x) %>%
#   rename(Hash = Sequence, Hash.y = Hash)

# Import the metadata file. Check to see if the order of your samples in the ASV table matches the order in the metadata table.
metadata <- read.csv(here("data", "baja_edna_metadata.csv"), row.names=1)
metadata <- metadata %>%
  mutate(Diver = recode(Diver,
                           "D4 (A)" = "D4",
                           "D4 (A) " = "D4",
                           "D4 (B)" = "D4"))
```

```{r remove samples with low sample sums}
asvs <- asvs %>%
  select(-BAJA_64, -BAJA_99, -BAJA_61, -BAJA_66)
```

```{r phyloseq format}
# Format the ASV and tax tables for phyloseq

# In this case each ASV (taxon) is a row
ASV = otu_table(asvs, taxa_are_rows=TRUE)

# For phyloseq taxonomy, we need to merge the tax and hash tables so that we have a list of hashes with associated taxonomy. 

# Now we can merge on the 'Sequence' column and then drop it from the merged data frame
tax_ps <- left_join(hash, tax, by='Sequence')
tax_ps <- select(tax_ps, -Sequence)

# Convert the hashes to row names and convert the whole data frame to a matrix
rownames(tax_ps) <- tax_ps$Hash
tax_ps <- select(tax_ps, -Hash)
tax_ps_mat <- as.matrix(tax_ps)

# Convert the taxonmy table into phyloseq format
TAX <- tax_table(tax_ps_mat)
```

```{r metadata}
# The Baja metadata row names are not the same as the ASV table column names, so edit to match
metadata <- metadata[order(rownames(metadata)), ]

# If you removed any samples during DADA2 processing, they need to be removed from the metadata file too
# Check the number of samples in the ASV table
length(colnames(asvs))
# Check the number of samples in the metadata table
length(rownames(metadata))

# They are the same.
```

```{r filter metadata}
# Make a list of the names we want to keep
samples.to.keep <- colnames(asvs)
metadata.filt <-filter(metadata, row.names(metadata) %in%
                                     samples.to.keep)

# Check again that the number of metadata rows matches the number of ASV table columns
length(colnames(asvs))
length(rownames(metadata.filt))

# Convert the metadata table into phyloseq format (use "metadata.filt" instead of "metadata" if you filtered out metadata rows)
MET <- sample_data(metadata)
```

```{r phyloseq prep}
# Make the phyloseq object from the ASV table, taxonomy table, and metadata table
ps <- phyloseq(ASV, MET, TAX)
```

```{r rarefy data}
# Calculate rarefaction curves
rarecurve(t(asvs), step = 100, sample = min(sample_sums(ps)), col = "blue", cex = 0.6)
min(sample_sums(ps))

# Rarefy to the minimum library size
psr <- rarefy_even_depth(ps, rngseed = 123, sample.size = min(sample_sums(ps)), replace = FALSE)

## Rarefy to 4000 reads (to see if too much diversity was removed by using the min)
#psr <- rarefy_even_depth(ps, rngseed = 123, sample.size = 4000, replace = FALSE)

# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- transform_sample_counts(psr, function(otu) otu/sum(otu))

# The sample_data has sample names as rownames, but we can add a column with the names as well
sample_data(ps.prop)['sample.id'] <-
  row.names(sample_data(ps.prop))
```

```{r phyloseq bar plots with contaminants}
# Here I am grouping by site
top20 <- names(sort(taxa_sums(ps.prop),
                    decreasing=TRUE))[1:20]

ps.top20 <- prune_taxa(top20, ps.prop)

bar <- plot_bar(ps.top20, x="Site_ID", fill="Family") +
  facet_wrap(~lat_group, scales="free_x")

bar

# Uh oh, looks like we have some non-fish families in here! Filter out these taxa and re-run the analyses.
```

```{r phyloseq filtertax}
# Let's filter out the contaminant families
ps.prop.pruned <- subset_taxa(ps.prop,
                                Family != "Bovidae")
ps.prop.pruned <- subset_taxa(ps.prop.pruned,
                                Family != "Hominidae")
ps.prop.pruned <- subset_taxa(ps.prop.pruned,
                                Family != "Suidae")
ps.prop.pruned <- subset_taxa(ps.prop.pruned,
                                Family != "Felidae")

# You may have empty samples after removing the contaminants. Remove those empty samples from the data set.
ps.prop.pruned.v2 <- prune_samples(sample_sums(ps.prop.pruned)>0, 
                                   ps.prop.pruned)

# Top 35 because that's how many hashes are needed to get to top 20 families
top35 <- names(sort(taxa_sums(ps.prop.pruned.v2),
                    decreasing=TRUE))[1:35]

ps.top35 <- prune_taxa(top35, ps.prop.pruned.v2)

bar2 <- plot_bar(ps.top35, x="Site_ID", fill="Family") +
  facet_wrap(~lat_group, scales="free_x")

bar2
```

```{r alpha diversity}
# Plot alpha diversity metrics for lat_group
alpha_lat <- plot_richness(ps.prop.pruned.v2, 
              x="lat_group", 
              measures=c("Simpson", "Shannon"),  
              color="lat_group") +
       labs(x = "Latitude Group",         
       color = "Latitude Group") +
      scale_color_discrete(labels = c("South", "Middle", "North")) +
      scale_x_discrete(labels = c("South", "Middle", "North")) +
      theme(axis.text.x = element_text(angle = 0))
alpha_lat

# Plot alpha diversity metrics for habitat
alpha_hab <- plot_richness(ps.prop.pruned.v2, 
              x="Most_Common_Habitat", 
              measures=c("Simpson", "Shannon"),  
              color="Most_Common_Habitat") +
              labs(x = "Habitat",         
              color = "Habitat") +
              theme(axis.text.x = element_text(angle = 0))
alpha_hab

# Plot alpha diversity metrics for max depth
alpha_dep <- plot_richness(ps.prop.pruned.v2, 
              x="MaxDepth", 
              measures=c("Simpson", "Shannon"),  
              color="MaxDepth") +
              labs(x = "Max Depth",
              color = "Max Depth") 
alpha_dep

# Plot alpha diversity metrics for diver
alpha_div <- plot_richness(ps.prop.pruned.v2, 
              x="Diver", 
              measures=c("Simpson", "Shannon"),  
              color="Diver")
alpha_div

# Calculate averages and do ANOVAs to compare
# Latitude group
# Extract Shannon and Simpson values
shannon_values_lat <- alpha_lat$data$value[alpha_lat$data$variable == "Shannon"]
simpson_values_lat <- alpha_lat$data$value[alpha_lat$data$variable == "Simpson"]

# Extract lat_group values for grouping
lat_group <- alpha_lat$data$lat_group[alpha_lat$data$variable == "Simpson"]

# Calculate the average for each lat_group
average_shannon_lat <- aggregate(shannon_values_lat ~ lat_group, FUN = mean)
average_simpson_lat <- aggregate(simpson_values_lat ~ lat_group, FUN = mean)

# Print the results
average_shannon_lat
average_simpson_lat

# Run ANOVA for Simpson
anova_result_lat <- aov(simpson_values_lat ~ lat_group)
summary(anova_result_lat)

# Habitat
# Extract Shannon and Simpson values
shannon_values_hab <- alpha_hab$data$value[alpha_hab$data$variable == "Shannon"]
simpson_values_hab <- alpha_hab$data$value[alpha_hab$data$variable == "Simpson"]

# Extract habitat values for grouping
habitat <- alpha_hab$data$Most_Common_Habitat[alpha_hab$data$variable == "Simpson"]

# Calculate the average for each habitat
average_shannon_hab <- aggregate(shannon_values_hab ~ habitat, FUN = mean)
average_simpson_hab <- aggregate(simpson_values_hab ~ habitat, FUN = mean)

# Print the results
average_shannon_hab
average_simpson_hab

# Run ANOVA for Simpson
anova_result_hab <- aov(simpson_values_hab ~ habitat)
summary(anova_result_hab)

# Max Depth
# Extract Shannon and Simpson values
shannon_values_dep <- alpha_dep$data$value[alpha_dep$data$variable == "Shannon"]
simpson_values_dep <- alpha_dep$data$value[alpha_dep$data$variable == "Simpson"]

# Extract depth values for grouping
depth <- alpha_dep$data$MaxDepth[alpha_dep$data$variable == "Simpson"]

# Calculate the average for each depth
average_shannon_dep <- aggregate(shannon_values_dep ~ depth, FUN = mean)
average_simpson_dep <- aggregate(simpson_values_dep ~ depth, FUN = mean)

# Print the results
average_shannon_dep
average_simpson_dep

# Run ANOVA for Simpson
anova_result_dep <- aov(simpson_values_dep ~ depth)
summary(anova_result_dep)

# Diver
# Extract Shannon and Simpson values
shannon_values_div <- alpha_div$data$value[alpha_div$data$variable == "Shannon"]
simpson_values_div <- alpha_div$data$value[alpha_div$data$variable == "Simpson"]

# Extract diver values for grouping
diver <- alpha_div$data$Diver[alpha_div$data$variable == "Simpson"]

# Calculate the average for each diver
average_shannon_div <- aggregate(shannon_values_div ~ diver, FUN = mean)
average_simpson_div <- aggregate(simpson_values_div ~ diver, FUN = mean)

# Print the results
average_shannon_div
average_simpson_div

# Run ANOVA for Simpson
anova_result_div <- aov(simpson_values_div ~ diver)
summary(anova_result_div)
```

```{r standardize}
# Hellinger transformation
# vegan equivalent: decostand
ps.hell <- transform_sample_counts(ps, function(x) sqrt(x / sum(x)))
ps.hell.pruned <- subset_taxa(ps.hell, Family != "Bovidae")
ps.hell.pruned <- subset_taxa(ps.hell.pruned, Family != "Hominidae")
ps.hell.pruned <- subset_taxa(ps.hell.pruned, Family != "Suidae")
ps.hell.pruned <- subset_taxa(ps.hell.pruned, Family != "Felidae")
ps.hell.pruned.v2 <- prune_samples(sample_sums(ps.hell.pruned)>0, 
                                   ps.hell.pruned)
```

```{r phyloseq NMDS hell}
# Plot beta diversity in NMDS

# Generate the ordinations using the Bray-Curtis distance metric
ord.nmds.bray <- ordinate(ps.hell.pruned.v2, "NMDS", "bray")

# Plot!

p1 <- plot_ordination(ps.hell, 
                      ord.nmds.bray, 
                      color="lat_group", 
                      title="PEDS NMDS Lat Group") +
  scale_color_discrete(name = "Lat Group", labels = c("South", "Middle", "North")) +
  #theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5))
p1
ggsave("beta_lat_peds.png", path = here("data", "figures"))

# Habitat
p2 <- plot_ordination(ps.hell, 
                      ord.nmds.bray, 
                      color="Most_Common_Habitat", 
                      title="PEDS NMDS Habitat") +
  #theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(color = "Habitat")
p2
#ggsave("beta_hab_peds.png", path = here("data", "figures"))

# Max Depth
p3 <- plot_ordination(ps.hell, 
                      ord.nmds.bray, 
                      color="MaxDepth", 
                      title="PEDS NMDS Max Depth") +
  #theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(color = "MaxDepth")
p3
#ggsave("beta_dep_peds.png", path = here("data", "figures"))

# Diver
p4 <- plot_ordination(ps.hell, 
                      ord.nmds.bray, 
                      color="Diver", 
                      title="PEDS NMDS Diver") +
  #theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(color = "Diver")
p4
#ggsave("beta_div_peds.png", path = here("data", "figures"))
```

```{r permanova hell}
# Extract the distance matrix (same one we used for the PCoA)
dist <- phyloseq::distance(ps.hell, "bray")
ps.perm.data <- data.frame(sample_data(ps.hell))

# Lat group
lat_group <- ps.perm.data$lat_group

# Test for the significance of season on sample distances
permanova <- adonis2(dist ~ lat_group,
                     ps.perm.data,
                     permutations=999)
permanova

# Habitat
Habitat <- ps.perm.data$Most_Common_Habitat

# Test for the significance of season on sample distances
permanova <- adonis2(dist ~ Habitat,
                     ps.perm.data,
                     permutations=999)
permanova

# Max Depth
MaxDepth <- ps.perm.data$MaxDepth

# Test for the significance of season on sample distances
permanova <- adonis2(dist ~ MaxDepth,
                     ps.perm.data,
                     permutations=999)
permanova

# Diver
Diver <- ps.perm.data$Diver

# Test for the significance of season on sample distances
permanova <- adonis2(dist ~ Diver,
                     ps.perm.data,
                     permutations=999)
permanova

# Multifactorial PERMANOVAs
permanova.v2 <- adonis2(dist ~ lat_group * Habitat,
                     ps.perm.data,
                     permutations=999)
permanova.v2

permanova.v2 <- adonis2(dist ~ lat_group * MaxDepth,
                     ps.perm.data,
                     permutations=999)
permanova.v2

permanova.v2 <- adonis2(dist ~ Habitat * MaxDepth,
                     ps.perm.data,
                     permutations=999)
permanova.v2
```
